src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -995,42 +999,42 @@ export default class Blockchain extends Emittery.Typed<
       }
     };
 
-    let txHashCurrentlyProcessing: string = null;
-    const transactionHash = Data.from(transaction.hash()).toString();
+    const transactionHash = transaction.hash();
+    let txHashCurrentlyProcessing: Buffer = null;
 
-    const beforeTxListener = async (tx: Transaction) => {
-      txHashCurrentlyProcessing = Data.from(tx.hash()).toString();
-      if (txHashCurrentlyProcessing == transactionHash) {
+    const beforeTxListener = (tx: Transaction) => {
+      txHashCurrentlyProcessing = tx.hash();
+      if (txHashCurrentlyProcessing.equals(transactionHash)) {
         if (keys && contractAddress) {
-          keys.forEach(async key => {
-            // get the raw key using the hashed key
-            let rawKey = await this.#database.storageKeys.get(
-              Data.from(key).toString()
-            );
-
-            vm.stateManager.getContractStorage(
-              Address.from(contractAddress).toBuffer(),
-              rawKey,
-              (err: Error, result: Buffer) => {
-                if (err) {
-                  throw err;
+          const database = this.#database;
+          return Promise.all(
+            keys.map(async key => {
+              // get the raw key using the hashed key
+              const rawKey: Buffer = await database.storageKeys.get(key);
```

Comment:
I've changed the way the keys are stored so we don't need to convert to a string any more.